### PR TITLE
rustdoc: add line breaks to where clauses a la rustfmt

### PR DESCRIPTION
**Pull request by QuietMisdreavus**
Much like my last PR for rustdoc (#36679), this adds line breaks to certain statements based on their line length. Here the focus was on where clauses.

Some examples:
- [Where clause in a trait function](https://shiva.icesoldier.me/custom-std/std/iter/trait.Iterator.html?search=#method.unzip) (also in the trait header block at the top of the page)
- [Where clause on a bare function](https://shiva.icesoldier.me/doc-custom2/petgraph/visit/fn.depth_first_search.html)
- [Where clauses in trait impls on a struct](https://shiva.icesoldier.me/custom-std/std/collections/struct.HashMap.html) (scroll to the bottom) These are regularly not on their own line, but will be given their own line now if their "prefix text" doesn't give them enough room to sensibly print their constraints. HashMap's trait impls provide some examples of both behaviors.

The libstd links above are the whole docs rendered with this, and the "bare function" link above is in another set that pulls some notable crates together. `petgraph` was the one that brought this request up, and that collection also includes [itertools](https://shiva.icesoldier.me/doc-custom2/itertools/trait.Itertools.html) which provided an easy sample to test with.

r? @GuillaumeGomez

### CODE REVIEW

File: src/librustdoc/html/format.rs [hunk 1]
```diff
@@ -50,7 +50,7 @@ pub struct MutableSpace(pub clean::Mutability);
 #[derive(Copy, Clone)]
 pub struct RawMutableSpace(pub clean::Mutability);
 /// Wrapper struct for emitting a where clause from Generics.
-pub struct WhereClause<'a>(pub &'a clean::Generics);
+pub struct WhereClause<'a>(pub &'a clean::Generics, pub String);
```
Comment (frewsxcv) [inline src/librustdoc/html/format.rs]: I'd be in favor of making this:

```rust
pub struct WhereClause<'a> {
    pub generics: &'a clean::Generics,
    pub pad: String,
};
```

So we're explicit about what the fields are instead of a collection of unnamed data structures.

File: src/librustdoc/html/format.rs [hunk 2]
```diff
@@ -718,30 +735,44 @@ impl fmt::Display for clean::Type {
...
 
 fn fmt_impl(i: &clean::Impl, f: &mut fmt::Formatter, link_trait: bool) -> fmt::Result {
+    let mut plain = String::new();
+
     if f.alternate() {
         write!(f, "impl{:#} ", i.generics)?;
     } else {
         write!(f, "impl{} ", i.generics)?;
     }
+    plain.push_str(&format!("impl{:#} ", i.generics));
```
Comment (frewsxcv) [inline src/librustdoc/html/format.rs]: Could this just be:

```rust
let mut plain = format!("impl{:#} ", i.generics);
```

...and then getting rid of the `String::new()` above?

Comment (QuietMisdreavus) [inline src/librustdoc/html/format.rs]: I don't see why not. I guess my intent was to have the declaration separate since I was going to be pushing more things onto it later.

File: src/librustdoc/html/format.rs [hunk 3]
```diff
@@ -157,52 +157,69 @@ impl fmt::Display for clean::Generics {
...
+            clause.push_str(" <span class='where'>where ");
         }
         for (i, pred) in gens.where_predicates.iter().enumerate() {
             if i > 0 {
-                f.write_str(", ")?;
+                if f.alternate() {
+                    clause.push_str(", ");
+                } else {
+                    clause.push_str(",<br>");
+                }
```
Comment (QuietMisdreavus) [inline src/librustdoc/html/format.rs]: Personal preference, I guess. It it part of a broader style guideline somewhere? At that point it's a ternary expression, too, and can be collapsed onto one line.

File: src/librustdoc/html/format.rs (hunk 1 above)
Comment (QuietMisdreavus) [inline src/librustdoc/html/format.rs]: No big deal to me either way. Looking at it again now, I kinda want to change that String (and Method's str, above) to integers since all they're conveying is the number of spaces to prepend onto new lines. They're passed differently, too, but handled the same.

**Comment by bluss** [comment]
Great work! It's an improvement for sure.

Maybe where clauses that are line broken should always be hanging on impls? Here's an image to illustrate http://i.imgur.com/UjP8aK9.png  (2) looks good and (1) is not as good.

File: src/librustdoc/html/format.rs [hunk 4]
```diff
@@ -157,52 +157,69 @@ impl fmt::Display for clean::Generics {
...
             }
             match pred {
                 &clean::WherePredicate::BoundPredicate { ref ty, ref bounds } => {
                     let bounds = bounds;
                     if f.alternate() {
-                        write!(f, "{:#}: {:#}", ty, TyParamBounds(bounds))?;
+                        clause.push_str(&format!("{:#}: {:#}", ty, TyParamBounds(bounds)));
```
Comment (ollie27) [inline src/librustdoc/html/format.rs]: It would be better to use `write!` instead of `.push_str(&format!(` to avoid an extra string allocation. For example, by importing `std::fmt::Write` this can be `write!(clause, "{:#}: {:#}", ty, TyParamBounds(bounds)).unwrap();`.

**Comment by QuietMisdreavus** [comment]
@bluss To be honest, minutes after I posted this PR I thought about whether that was the way to go. Doing that change would be pretty easy - I'll go ahead and work that in.

**Comment by bluss** [comment]
Here's a way to count the length of a formatting without making the string. https://is.gd/Yks7tx

File: src/librustdoc/html/render.rs [hunk 5]
```diff
@@ -2899,10 +2945,12 @@ fn render_impl(w: &mut fmt::Formatter, cx: &Context, i: &Impl, link: AssocItemLi
 
 fn item_typedef(w: &mut fmt::Formatter, cx: &Context, it: &clean::Item,
                 t: &clean::Typedef) -> fmt::Result {
+    let indent = format!("type {}{:#} ", it.name.as_ref().unwrap(), t.generics);
+    let indent = repeat(" ").take(indent.len()).collect::<String>();
```
Comment (QuietMisdreavus) [inline src/librustdoc/html/render.rs]: Oh yeah, I remember you talking about that on IRC. This is a pattern used more broadly than my changes, so I'd be more comfortable taking that to a separate PR? I dunno what the general consensus on that would be, though.

**Comment by QuietMisdreavus** [comment]
@bluss On that counting code: Looking at it again, it seems to be just a byte count? That might backfire if non-ascii idents are allowed. On the other hand, I've already been falling into the same trap by using `len()` myself, so it's not a substantial change from what I've already been doing. I'd be willing to work it in.

**Comment by bluss** [comment]
Yes that's true. It's the same count as `String::len`'s, though.

**Comment by bluss** [comment]
The result looks great, and I'd really like this. I think it would be a lot easier to accept if we had a cleaner way to do the plain/vs regular formatting, without the duplicate code for each piece.

**Comment by alexcrichton** [comment]
@rust-lang/docs thoughts about this PR? It seems to be getting a little old, but it's still mergeable! Would nice to get a resolution!

**Comment by GuillaumeGomez** [comment]
@alexcrichton: I totally forgot about this one. Since people are globally in favor for it, let's merge it. It's still possible in the future to remove it anyway.

@QuietMisdreavus: Thanks for your work and sorry for the delay!

@bors: r+ rollup
